File: src/ros2_bridge/mappers/nav_msgs/grid_cells.rs
//! Typed mapper for `nav_msgs/msg/GridCells`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn grid_cells_to_bus(msg: ros_env::nav_msgs::msg::GridCells) -> crate::nav_msgs::msg::v1::GridCells {
    crate::nav_msgs::msg::v1::GridCells {
        header: Some(crate::ros2_bridge::mappers::std_msgs::header::header_to_bus(msg.header)),
        cell_width: msg.cell_width,
        cell_height: msg.cell_height,
        cells: msg.cells.into_iter().map(crate::ros2_bridge::mappers::geometry_msgs::point::point_to_bus).collect(),
    }
}

pub(crate) fn grid_cells_to_ros(bus: crate::nav_msgs::msg::v1::GridCells) -> ros_env::nav_msgs::msg::GridCells {
    ros_env::nav_msgs::msg::GridCells {
        header: crate::ros2_bridge::mappers::std_msgs::header::header_to_ros(bus.header.unwrap_or_default()),
        cell_width: bus.cell_width,
        cell_height: bus.cell_height,
        cells: bus.cells.into_iter().map(crate::ros2_bridge::mappers::geometry_msgs::point::point_to_ros).collect(),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct NavMsgsGridCellsMapper;

impl TypedTopicMapper for NavMsgsGridCellsMapper {
    type Ros = ros_env::nav_msgs::msg::GridCells;
    type Bus = crate::nav_msgs::msg::v1::GridCells;

    fn type_name(&self) -> &'static str {
        "nav_msgs/msg/GridCells"
    }

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(grid_cells_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(grid_cells_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/nav_msgs/grid_cells.py
"""Generated mapper for `nav_msgs/msg/GridCells`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.std_msgs.header import header_to_bus, header_to_ros
from robot_bus.ros2_bridge.mappers.geometry_msgs.point import point_to_bus, point_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.nav_msgs.msg.v1 import GridCells as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def grid_cells_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.header.CopyFrom(header_to_bus(msg.header))
    bus.cell_width = msg.cell_width
    bus.cell_height = msg.cell_height
    bus.cells.extend([point_to_bus(x) for x in msg.cells])
    return bus


def grid_cells_to_ros(bus):
    from nav_msgs.msg import GridCells as RosMsg

    out = RosMsg()
    out.header = header_to_ros(bus.header)
    out.cell_width = bus.cell_width
    out.cell_height = bus.cell_height
    out.cells = [point_to_ros(x) for x in bus.cells]
    return out


class NavMsgsGridCellsMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from nav_msgs.msg import GridCells as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return grid_cells_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return grid_cells_to_ros(bus)
